src/reader.rs

```diff
@@ -4796,35 +4796,64 @@
                 tokens.first().map(|tok| tok.offset()).unwrap_or(cursor_pos),
             ) == search_string_range
         };
+
+        // Only to be used if no case-sensitive suggestions are found.
+        let mut icase_history_result = None;
+
         if cursor_line_has_process_start {
-            let mut searcher = HistorySearch::new_with_type(
+            let mut searcher = HistorySearch::new_with(
                 history,
                 search_string.to_owned(),
                 SearchType::LinePrefix,
+                SearchFlags::IGNORE_CASE,
+                0,
             );
+
             while !ctx.check_cancel() && searcher.go_to_next_match(SearchDirection::Backward) {
                 let item = searcher.current_item();
 
-                // Suggest only a single line each time.
-                let matched_line = item
+                // The history item's may have multiple lines of text.
+                // Only suggest the line that actually contains the search string.
+
+                let lines = item
                     .str()
                     .as_char_slice()
                     .split(|&c| c == '\n')
                     .rev()
-                    .find(|line| line.starts_with(search_string.as_char_slice()))
-                    .unwrap();
+                    .map(wstr::from_char_slice);
+
+                let mut icase = false;
+                let mut matched_line = lines.clone().find(|line| line.starts_with(search_string));
+
+                // Only check for a case-insensitive match if we haven't already found one
+                if matched_line.is_none() && icase_history_result.is_none() {
+                    icase = true;
+                    matched_line = lines
+                        .into_iter()
+                        .find(|line| string_prefixes_string_case_insensitive(search_string, line));
```

Comment:
It's kind of a shame that we have to match again, just to find
out the match quality.  I guess that's owed to the weirdness of
SearchType::LinePrefix, I'm not sure if there's a nice way to return
this information to the caller. Fine.